Added document for initial nullability API design. (#1)
333fred/roslyn: docs/compilers/CSharp/Nullability Public API Design Notes.md

# Nullability API Design Notes

## Guiding Principles

> When designing this API, there were a few core principles that we tried to keep in mind.

Reviewer: ```suggestion
When designing this API, there were a few core principles that we tried to keep in mind:
``` #Closed

> 1. First and foremost, we do not want to break existing analyzers and refactorings. This means that common patterns that used to work need to continue to work. These include reference equality for `ITypeSymbol`s that aren't generics, such as `String`.

Reviewer: ```suggestion
1. First and foremost, we do not want to break existing analyzers and refactorings. This means that common patterns that used to work need to continue to work. These include reference equality for `ITypeSymbol`s that aren't generics, such as `string`.
``` #Closed

> 1. First and foremost, we do not want to break existing analyzers and refactorings. This means that common patterns that used to work need to continue to work. These include reference equality for `ITypeSymbol`s that aren't generics, such as `String`.

Reviewer: ```suggestion
1. First and foremost, we do not want to break existing analyzers and refactorings. This means that common patterns that used to work need to continue to work. These include reference equality for `ITypeSymbol`s that aren't generic, such as `String`.
``` #Closed

> 1. First and foremost, we do not want to break existing analyzers and refactorings. This means that common patterns that used to work need to continue to work. These include reference equality for `ITypeSymbol`s that aren't generics, such as `String`.

Reviewer 2: >1. First and foremost [](start = 0, length = 21)

Please consider breaking long lines. #Closed

> 2. We'd like the ability to avoid computing nullability information if it's not required. Currently, the SemanticModel APIs do a good job of only causing the minimal amount of binding required to answer the question being asked, but attempting to compute nullability information will cause significantly more to need to be bound. To fully determine nullability, we must bind the entire method and run nullability analysis up to the point requested by the caller, unlike today where individual statements can get away with being bound by themselves. Even if the standard path calculates nullability info, we feel that some scenarios will be both perf sensitive and not care about results of nullability analysis. This includes things like intellisense.

Reviewer: Consider "Semantic Model" (two words) or "`SemanticModel`" #Closed

> 2. We'd like the ability to avoid computing nullability information if it's not required. Currently, the SemanticModel APIs do a good job of only causing the minimal amount of binding required to answer the question being asked, but attempting to compute nullability information will cause significantly more to need to be bound. To fully determine nullability, we must bind the entire method and run nullability analysis up to the point requested by the caller, unlike today where individual statements can get away with being bound by themselves. Even if the standard path calculates nullability info, we feel that some scenarios will be both perf sensitive and not care about results of nullability analysis. This includes things like intellisense.

Reviewer: ```suggestion
2. We'd like the ability to avoid computing nullability information if it's not required. Currently, the SemanticModel APIs do a good job of only causing the minimal amount of binding required to answer the question being asked, but attempting to compute nullability information will cause significantly more to need to be bound. To fully determine nullability, we must bind the entire method and run nullability analysis up to the point requested by the caller, unlike today where individual statements can get away with being bound by themselves. Even if the standard path calculates nullability info, we feel that some scenarios will be both perf sensitive and not care about results of nullability analysis. This includes features like intellisense.
``` #Closed

> 2. We'd like the ability to avoid computing nullability information if it's not required. Currently, the SemanticModel APIs do a good job of only causing the minimal amount of binding required to answer the question being asked, but attempting to compute nullability information will cause significantly more to need to be bound. To fully determine nullability, we must bind the entire method and run nullability analysis up to the point requested by the caller, unlike today where individual statements can get away with being bound by themselves. Even if the standard path calculates nullability info, we feel that some scenarios will be both perf sensitive and not care about results of nullability analysis. This includes things like intellisense.

Reviewer 2: >to need [](start = 309, length = 7)

`to need` can be deleted. #Closed

> 2. We'd like the ability to avoid computing nullability information if it's not required. Currently, the SemanticModel APIs do a good job of only causing the minimal amount of binding required to answer the question being asked, but attempting to compute nullability information will cause significantly more to need to be bound. To fully determine nullability, we must bind the entire method and run nullability analysis up to the point requested by the caller, unlike today where individual statements can get away with being bound by themselves. Even if the standard path calculates nullability info, we feel that some scenarios will be both perf sensitive and not care about results of nullability analysis. This includes things like intellisense.

Reviewer 2: >things like intellisense [](start = 726, length = 24)

`... scenarios like Intellisense` or `... IntelliSense.` #Closed

> 2. We'd like the ability to avoid computing nullability information if it's not required. Currently, the SemanticModel APIs do a good job of only causing the minimal amount of binding required to answer the question being asked, but attempting to compute nullability information will cause significantly more to need to be bound. To fully determine nullability, we must bind the entire method and run nullability analysis up to the point requested by the caller, unlike today where individual statements can get away with being bound by themselves. Even if the standard path calculates nullability info, we feel that some scenarios will be both perf sensitive and not care about results of nullability analysis. This includes things like intellisense.

Reviewer 3: ```suggestion
2. We'd like the ability to avoid computing nullability information if it's not required. Currently, the SemanticModel APIs do a good job of only causing the minimal amount of binding required to answer the question being asked, but attempting to compute nullability information will cause significantly more to need to be bound. To fully determine nullability, we must bind the entire method and run nullability analysis up to the point requested by the caller, unlike today where individual statements can get away with being bound by themselves. Even if the standard path calculates nullability info, we feel that some scenarios will be both perf-sensitive and not care about results of nullability analysis. This includes things like completion.
``` #Closed

> 3. Users should be able to get at both declared nullability for variables, as well as seeing the current flow state of expressions based on flow analysis.

Reviewer 2: >seeing the current flow state of expressions based on flow analysis [](start = 86, length = 67)

Perhaps `the inferred nullability at each expression from flow analysis`. (`current` seems to imply a different result at different times.) #Closed

## General Concepts

> Nullability information is exposed publicly as an enum, `Nullability`. There are 5 possible values:

Reviewer 4: >5 possible values [](start = 81, length = 17)

If it's not covered below, it would be good to clarify what enum value is used for:
- a value type
- a nullable value type
- various generic type parameter scenarios #Closed

* `NotApplicable` - This is used when nullability information is requested about something that does not have a nullability, such as a statement, directive, or other similar constructs.
* `NotAnalyzed` - This is used to fulfill priority #2. If the user requests information while asking for nullability to not be calculated, we'll use this to stand in for nullability.

> * `Unknown` - This is used for legacy C# code that does not know about nullability. This can appear in modern C# code that calls unannoted C#. For use in this document, we represent unknown via `~`, but this symbol does not appear in source.

Reviewer 3: ```suggestion
* `Unknown` - This is used for legacy C# code that does not know about nullability. This can appear in modern C# code that calls unannotated C#. For use in this document, we represent unknown via `~`, but this character does not appear in source.
``` #Closed

Reviewer 3: ("symbol does not appear in source" implies to me that symbols won't have this) #Closed

* `NotNull` - This is used when we know the nullability is not null.

> * `MaybeNull` - This is used when we know the nullability might be null.

Reviewer 3: Did we debate "may be null" or "maybe null"? "MaybeNull means it may be null" implies it the b should be capitalized? #Closed

> 

Reviewer 5: I think this set of values is still under discussion, particularly how to represent an unconstrained type parameter. #Closed

Author: Added a note later that this is still being debated.

---
In reply to: [239620678](https://github.com/333fred/roslyn/pull/1#discussion_r239620678) [](ancestors = 239620678)

> Nullability information will be retreived from the `SemanticModel`, much like type information. `GetSemanticModel` will return a `SemanticModel` that is aware of nullability information, and will force binding and calculation of nullability information when queried. We will offer a method off of `SemanticModel`, `GetSingleStatementSemanticModel()`, which will return a `SemanticModel` that does not calculate nullability information, for particularly performance-sensitive scenarios. This `SemanticModel` will share a cache with its parent, much like speculative `SemanticModel`s, and all nullabilities accessed on symbols retrieved from this `SemanticModel` will be `NotAnalyzed`.

Reviewer 4: >retreived [](start = 32, length = 9)

typo: retrieved #Closed

> Nullability information will be retreived from the `SemanticModel`, much like type information. `GetSemanticModel` will return a `SemanticModel` that is aware of nullability information, and will force binding and calculation of nullability information when queried. We will offer a method off of `SemanticModel`, `GetSingleStatementSemanticModel()`, which will return a `SemanticModel` that does not calculate nullability information, for particularly performance-sensitive scenarios. This `SemanticModel` will share a cache with its parent, much like speculative `SemanticModel`s, and all nullabilities accessed on symbols retrieved from this `SemanticModel` will be `NotAnalyzed`.

Reviewer 3: How are we documenting GetSingleStatementSemanticModel? Is it strictly stated "it just doesn't compute nullability" or are we leaving the door open for other flow things like this down the road? #Closed

Author: No longer relevant.

---
In reply to: [240430403](https://github.com/333fred/roslyn/pull/1#discussion_r240430403) [](ancestors = 240430403)

Generally speaking, when you want to determine the nullability of an `ITypeSymbol`, you need to check the containing context. The declaring symbol (e.g. `ILocalSymbol`, `IFieldSymbol`, etc.) will have the declared nullability of the type. `GetTypeInfo` will give you both the `ITypeSymbol` and the `Nullability` of an expression. Nested nullabilities for type parameters are contained on the containing `ITypeSymbol`. For example, if you wanted to know whether an array's contents are nullable, you will need to check the `IArrayTypeSymbol` for `ElementNullability`.

### Declared Nullability

Declared nullability is the nullability that was explicitly typed in the source code by the programmer. There are some cases where this is inferred, such as lambda arguments/return types, `var` variables, and type parameter substitution. For these cases, flow state is necessary to compute the declared nullability. For example:

```C#
string? s1 = string.Empty;
var s2 = s1;
s2 = null; // This will warn, because the type of s2 is string, not string?
```

> The declared nullability of `s1` is `MaybeNull`, even though it is being initialized with a non null value. However, because of the flow state of `s1` is `NotNull`, we will infer the declared nullability of `s2` to be `NotNull` as well.

Reviewer 4: >non null [](start = 92, length = 8)

nit: non-null? #Closed

Author: It was originally `NonNull`, review decided it should be `NotNull`. #Closed

Reviewer 4: Sorry I wasn't clear. I was suggesting "...it is being initialized with a **non-null** value ..." #Closed

Declared nullability can be retreived from the declaration symbol (e.g. `IFieldSymbol`, `IMethodSymbol`, etc.).

### Flow State

> When the nullable feature is enabled, Roslyn will track the flow state of expressions throughout a method, regardless of what the variable was declared as. `SemanticModel` can be used to request flow state information for individual expressions or sub-expressions through the `GetTypeInfo` APIs. The `TypeInfo` struct will be augmented with `Nullability` and `ConvertedNullability` information, similar to how it currently has `Type` and `ConvertedType`. Nullability of symbols at a location can be retrieved through the `GetSymbolInfo` API: if the expression successfully resolved to a single symbol, the `Symbol` property will have nullability calculated. `CandidateSymbols` will not have calculated nullabilities, as we cannot run nullable analysis on code that has failed overload resolution.

Reviewer 3: Does that mean CandidateSymbols will still have declared accessibilities or will be marked with `NotAnalyzed` to explicitly state it wasn't analyzed? #Closed

Author: CandidateSymbols will likely be `NotAnalyzed`. If overload resolution did not succeed we cannot run nullability analysis.

---
In reply to: [240431601](https://github.com/333fred/roslyn/pull/1#discussion_r240431601) [](ancestors = 240431601)

> ### `GetSingleStatementSemanticModel`

Reviewer: Don't believe that backticks on a # line will have any effect. #Closed

Author: ![image](https://user-images.githubusercontent.com/2371880/49532267-faa5de00-f870-11e8-8f58-76dabda8761a.png)
 #Closed

Reviewer: Huh. Guess they changed that.  #Closed

> This API proposal changes the way `SemanticModel` calls behave by default. Currently, if semantic information is requested for a `SyntaxNode`, we make an effort to bind as little as possible to provide semantic information about that statement, usually a single statement. However, if that semantic information will include `Nullability`, we must bind all statements that come before that statement and run nullability analysis up to and including that statement. We believe that this change will be ok for most scenarios, as things such as colorization and analyzers are already running in the background near-constantly, binding pretty much everything on screen and in the file anyway. There are a few specific scenarios that will be perf-sensitive, and `GetSingleStatementSemanticModel()` will be provided for these. It will share an initial-binding `BoundNode` cache with the standard semantic model, so it will get the benefits if something else with the same semantic model requests information, as it would have today.

Reviewer 6: `GetSingleStatementSemanticModel()` seems a bit obscure as a name. At least, it seems scary for something that is effectively "give me a semantic model that does the same level of analysis as a normal semantic model used to". I would expect users to see this name and assume that it's supposed to be used in advanced scenarios, while the reality is that it is more perf-conscious. #Closed

Reviewer 7: > I would expect users to see this name and assume that it's supposed to be used in advanced scenarios, while the reality is that it is more perf-conscious.

Well, in .NET it seems we have (and often still) treat performance as an advanced scenario, if that means loss of fidelity or other shortcomings the consumer needs to be aware of. In your case, maybe you want to make it an overload that takes a parameter, like `GetSemanticModel(bool skipNullabilityInformation)`. The existing overload (logically) calls this with `false`. #Closed

Reviewer 6: @terrajobst: I totally agree that we treat performance as an advanced scenario, and we should. I think this case is _slightly_ different since we're making the normal code path less efficient to provide more correctness and offering a way to get back the original behavior, which happens to be more efficient. I think you're suggested overload is a good one. #Closed

Author: > I would expect users to see this name and assume that it's supposed to be used in advanced scenarios, while the reality is that it is more perf-conscious.

Yes, and I this was my intent. I thought that in the design meeting we had come to the conclusion that there were really only a few IDE scenarios that we expected to need this perf saving, and that everyone else really won't get much of an advantage from this API.

The problem with `GetSemanticModel(bool skipNullabiliytInformation)` is twofold:
1. The nullability-unaware and nullability-aware semantic models will be unable to share a bound node cache. This _would_ simplify implementation, but one of the intents here was that the nullable-aware model would share the initial bind cache of bound nodes with the unaware model.
2. One of my intents with with API was to provide a way to get a performant `SemanticModel`, without saying precisely what we do to get that performance. This way, if we add future analysis passes to the default model, we don't overwhelm the user with options for turning off individual things. #Closed

Reviewer 6: > The nullability-unaware and nullability-aware semantic models will be unable to share a bound node cache.

Why is that? If this is the right API shape, I would think we could adjust the implementation to make it work (`ConditionalWeakTable` may be your friend here :smile:).

> One of my intents with with API was to provide a way to get a performant SemanticModel, without saying precisely what we do to get that performance. This way, if we add future analysis passes to the default model, we don't overwhelm the user with options for turning off individual things.

I definitely agree with that principle, but the name `GetSingleStatementSemanticModel` is probably too specific to cover any possible future analysis passes as well. I think you could achieve your goal here by picking a different name for the flag (e.g. `performFullAnalysis` with a default of true). #Closed

Author: @jasonmal, what do you think of this proposal from an IDE usage perspective? Give Dustin's points I'd probably be fine with a `GetSemanticModel(bool performFullAnalysis)`, but I'd like to hear your thoughts on usage. #Closed

Reviewer 7: I suggest you negate the Boolean. Given that you have an existing method `GetSemanticModel()` that takes no arguments, you cannot indicate what the default value is. In the framework we try to say that the default value is `default`. In your case, that would mean something like `GetSemanticModel(bool simplifiedAnalysis)`.

When I proposed the original API I thought they way you'd evolve this is by letting the caller choose which part of the analysis they are willing to sacrifice by adding more overloads, like

```C#
// V1
GetSemanticModel();
// V2
GetSemanticModel(bool skipNullabiliytInformation);
// V3
GetSemanticModel(bool skipNullabiliytInformation, bool skipBorrowingInformation);
```

This way, `GetSemanticModel()` always returns the most information while you always have a specialized way to optimize for speed. #Closed

Author: @terrajobst while I can see some value in negating the boolean, I'm specifically avoiding the ability to configure multiple different analyses here. I don't think we intend this API for consumers to decide "oh, we're only going to need this part of the information so lets save on analysis time". My understanding of this API is that it's for _very_ specific scenarios that are extremely performance-constrained and only need information from initial binding. #Closed

Reviewer 3: I endorse @terrajobst's proposal, if only because it's also what I was thinking in the meetings all along. I imagine one of two futures:

1. We never have a second type of analysis that we want to disable like this. Great, the scenario you are worried about won't happen.
2. We have a second type of analysis that you want to disable. Unless you're willing to bet me $1,000,000 (bitcoin not accepted) that it will fall in the same classification as this type _and_ all the features that wanted or didn't want it are the exact same set as today...you're going to have somebody who wants to turn one and one off. At this point your API won't work at all, and you'll be probably doing what @terrajobst is arguing for.

> One of my intents with with API was to provide a way to get a performant SemanticModel, without saying precisely what we do to get that performance.

SemanticModels are already magic, but fundamentally writing any analysis means you need to know what's happening. If you made SemanticModel faster by limiting it's analysis and returning a best guess after a few milliseconds of CPU...we'd be at your door with pitchforks because you just broke all of our stuff.

I assume your concern 1 about bound node caches is no longer a concern? If you could implement GetSingleStatementSemanticModel then I don't see how (today) the implementation of this other API would just be would be just:

`GetSemanticModel(bool skipNullability) => skipNullability ? GetSingleStatementSemanticModel() : this`

Which doesn't introduce new complexity. #Closed

## New APIs

We add the new APIs for retreiving and interacting with nullability:

```C#

> public enum Nullability

Reviewer 4: >Nullability [](start = 12, length = 11)

missing an enum value? #Closed

> public enum Nullability

Reviewer 5: >public [](start = 0, length = 6)

This contradicts the earlier list of these enum values. #Closed

{
    NotApplicable,
    NotComputed,
    NotNull,
    MaybeNull
}

#region Declared Nullability

> // These can potentially be influenced by calculated flow state, but do not change after initial declaration

Reviewer 3: You've broken "Nullability" into two sections in this spec. Is that something you imagine users will ever care about, or is this just for our understanding? As a consumer I'm not sure the distinction is meaningful; if you don't expect it's meaningful to me either, great. #Closed

Author: I don't imagine users will ever care about this in practice, but it's important to understand the concepts from an API design perspective.

---
In reply to: [240433882](https://github.com/333fred/roslyn/pull/1#discussion_r240433882) [](ancestors = 240433882)

public interface IDiscardSymbol : ISymbol
{
    Nullability Nullability { get; }
}

public interface IEventSymbol : ISymbol
{
    Nullability Nullability { get; }
}

public interface IFieldSymbol : ISymbol
{
    Nullability Nullability { get; }
}

public interface ILocalSymbol : ISymbol
{
    Nullability Nullability { get; }
}

public interface IMethodSymbol : ISymbol
{
    Nullability ReturnNullability { get; }

>     Nullability TypeArgumentsNullabilities { get; }

Reviewer 2: `ImmutableArray<Nullability>` #Closed

    Nullability ReceiverNullability { get; }
}

public interface IParameterSymbol : ISymbol
{
    Nullability Nullability { get; }
}

#endregion

#region Flow State Nullability

public interface IArrayTypeSymbol : ITypeSymbol
{
    Nullability ElementNullability { get; }
}

public interface INamedTypeSymbol : ITypeSymbol
{
    ImmutableArray<Nullability> TypeArgumentsNullabilities { get; }
}

public interface ITypeParameterSymbol : ITypeSymbol
{
    Nullability ReferenceTypeConstraintNullability { get; }

>     Nullability ConstraintsNullabilities { get; }

Reviewer 3: Array? #Closed

}

public struct TypeInfo
{
    public Nullability Nullability { get; }
    public Nullability ConvertedNullability { get; }
}

#endregion

#region SemanticModel

public abstract class SemanticModel
{
    public SemanticModel GetSingleStatementSemanticModel();
}

#endregion

#region Printing

public interface ITypeSymbol
{
    string ToDisplayString(Nullability nullability, SymbolDisplayFormat format = null);
    ImmutableArray<SymbolDisplayPart> ToDisplayParts(Nullability nullability, SymbolDisplayFormat format = null);
    string ToMinimalDisplayString(SemanticModel semanticModel, int position, Nullability nullability, SymbolDisplayFormat format = null);
    ImmutableArray<SymbolDisplayPart> ToMinimalDisplayString(SemanticModel semanticModel, int position, Nullability nullability, SymbolDisplayFormat format = null);
}

/*
 * There will be more API changes, such as on SyntaxFactory and SyntaxGenerator. This section will be updated later when we have designs for this ready.
 */

#endregion
```

> ## Alternatives

Reviewer 3: ```suggestion
## Considered Alternatives
```

(just to clarify yes, we did consider them!) But thanks for including this section, this illustrates we did go through this design. #Closed

### Nullability on ITypeSymbol

> The basic principle of this proposal is to add `Nullability Nullability { get; }` to `ITypeSymbol`, and remove the rest of the proposed nullability retrieval APIs in the main section. This has a few benefits:

Reviewer 3: Do you mean "this proposal" meaning the doc or this alternative?

Reviewer 3: (ditto for the rest)

For these reasons, we have decided not take this approach.

### GetNullabilityAwareSemanticModel

This proposal flipped the `GetSingleStatementSemanticModel` proposal on its head: instead of `SemanticModel` calculating nullability by default, it would only calculate when requested. We decided against this because we believe that, except for specific cases where perf is an issue, the savings gained from not calculating nullability would be moot as IDE features or analyzers will have already requested nullability anyway, and the time savings would be unused. Therefore, we are not taking this approach.

### `GetNullabilityInfo`

An early proposal had a parallel API to `GetTypeInfo()`, `GetNullabilityInfo()`, that would return a structure similar to `TypeInfo` that contained nullability information. We decided that this would provide no benefits: because of nested nullability, we would still need to run nullability analysis on calls to `GetTypeInfo`, and it would needlessly separate information that users will want to get access to in tandem. Additionally, we could not come up with a good set of APIs for `GetDeclaredSymbol` and `GetSymbolInfo` to go along with that API.

> ### Lazily calcualted `Nullability` properties

Reviewer 3: ```suggestion
### Lazily calculated `Nullability` properties
```
